workspace-manager-clienttests/src/main/java/scripts/uploadscripts/SummariesToBigQuery.java

```diff
@@ -0,0 +1,215 @@
+    protected MeasurementCollectionScript.MeasurementResultSummary[] measurementCollectionSummaries;
+
+    /**
+     * Setter for any parameters required by the upload script. These parameters will be set by the
+     * Result Uploader based on the current Upload List, and can be used by the upload script methods.
+     *
+     * @param parameters list of string parameters supplied by the upload list
+     */
+    public void setParameters(List<String> parameters) throws Exception {
+        if (parameters == null || parameters.size() < 2) {
+            throw new IllegalArgumentException(
+                    "Must provide BigQuery project_id and dataset_name in the parameters list");
+        }
+        projectId = parameters.get(0);
+        datasetName = parameters.get(1);
+    }
+
+    private static String testRunTableName = "testRun";
+    private static String testScriptResultsTableName = "testScriptResults";
+    private static String measurementCollectionTableName = "measurementCollection";
+
+    /**
+     * Upload the test results saved to the given directory. Results may include Test Runner
+     * client-side output and any relevant measurements collected.
+     */
+    public void uploadResults(
+            Path outputDirectory, ServiceAccountSpecification uploaderServiceAccount) throws Exception {
+        // get a BigQuery client object
+        logger.debug("BigQuery project_id:dataset_name: {}:{}", projectId, datasetName);
+        BigQuery bigQueryClient =
+                BigQueryUtils.getClientForServiceAccount(uploaderServiceAccount, projectId);
+        // read in TestConfiguration TestRunSummary, and array of
+        // MeasurementCollectionScript.MeasurementResultSummary objects
+        renderedTestConfiguration = TestRunner.getRenderedTestConfiguration(outputDirectory);
+        testRunSummary = TestRunner.getTestRunSummary(outputDirectory);
+        measurementCollectionSummaries =
+                MeasurementCollector.getMeasurementCollectionSummaries(outputDirectory);
+
+        // insert a single row into testRun
+        if (BigQueryUtils.checkRowExists(
+                bigQueryClient, projectId, datasetName, testRunTableName, "id", testRunSummary.id)) {
+            logger.info(
+                    "A row with this id already exists in the "
+                            + testRunTableName
+                            + " table. Inserting a duplicate.");
+        }
+        TableId tableId = TableId.of(datasetName, testRunTableName);
+        InsertAllRequest insertRequest =
+                InsertAllRequest.newBuilder(tableId).addRow(buildTestRunRow(outputDirectory)).build();
+        BigQueryUtils.insertAllIntoBigQuery(bigQueryClient, insertRequest);
+
+        // insert into testScriptResults
+        tableId = TableId.of(datasetName, testScriptResultsTableName);
+        InsertAllRequest.Builder insertRequestBuilder = InsertAllRequest.newBuilder(tableId);
```

Comment:
FYI: using streaming upload can result in duplicates. https://cloud.google.com/bigquery/streaming-data-into-bigquery#dataconsistency. you can add a unique `insertId` pseudo-column to each row to help.

Reply:
Added your suggestion to enhancement [doc](https://docs.google.com/document/d/1gYqlJTpC4C4S1K0pYikeNJx-HWNZAZJYybinmuqcSTI/edit#).

I agree, should use a `sequence` or `version` or combination of both to manage concurrent inserts or updates.